re-add correctly region-extracted cell measures and ancillary variables after `extract_region` (for Changelog v2.10: authors: @valeriupredoi and @schlunma)

valeriupredoi opened the pull request:
<!--
    Thank you for contributing to our project!

    Please do not delete this text completely, but read the text below and keep
    items that seem relevant. If in doubt, just keep everything and add your
    own text at the top, a reviewer will update the checklist for you.

-->

## Description

<!--
    Please describe your changes here, especially focusing on why this pull
    request makes ESMValCore better and what problem it solves.

    Before you start, please read our contribution guidelines: https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html

    Please fill in the GitHub issue that is closed by this pull request,
    e.g. Closes #1903
-->

Cube intersection removes all cell measures and ancillary variables in certain conditions (for certain regions ie certain coordinates of the region box). This has been raised upstream, see below. NOTE for Codacy fails: I intentionally placed the imports inside block for ease of removal when we will eventually remove the workaround code block,

Closes #2162 

Brought to my attention by @rebeccaherman1 in the above linked issue, which is, in fact a bug in iris - raised upstream in https://github.com/SciTools/iris/issues/5413 - this is a workaround (temporary too) so that we have something outputting correct data before iris fix it at their end.


***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

valeriupredoi committed:
readd cell measure after extract region

valeriupredoi committed:
add conditionals

valeriupredoi committed:
readd ancillary variable

valeriupredoi committed:
fix flake

valeriupredoi committed:
yapf the crap outta it

valeriupredoi committed:
can flake pls shut up

valeriupredoi committed:
add test case dir

valeriupredoi committed:
appease flake

valeriupredoi committed:
add test netcdf file

valeriupredoi committed:
polish inline comments

valeriupredoi commented:
BTW @ledm I have a feeling that this bit you as well, in your continual battle with fx variables

schlunma reviewed (changes requested):
Thanks V for tackling this!

I have three main comments:

1. I think you can simplify this by re-using `extract_region` on a cell measure/ancillary variable cube (see code suggestions).
2. Why do change the formatting of unrelated code? This has been formatted by `black` to look nice and shiny :star: 
3. Could you maybe come up with a test that doesn't require adding a 1.1 MB file to the repo? We have plenty of example code that sets up a cube with a `CellMeasure`, e.g., [here](https://github.com/ESMValGroup/ESMValCore/blob/3222d28a8587e7ce3f5871522c035079b645dd5b/tests/unit/preprocessor/_regrid/test_extract_levels.py#L84).

valeriupredoi commented on esmvalcore/preprocessor/_area.py:
```diff
@@ -87,6 +91,66 @@ def extract_region(cube, start_longitude, end_longitude, start_latitude,
+    # When removing this block after iris have a fix, make sure to remove the
+    # test too tests/integration/preprocessor/_extract_region/
+    # setp 1: cell measures
+    region_cell_meas = region_subset.cell_measures()
+    if (cell_measures, cube.coord('latitude').ndim
+            == 1) and not region_cell_meas:
```
the last check element is for the case when `extract_region` actually preserves the cell measures - it's a bizarre behaviour that sometimes, depending on the region's bounds, `extract_region` (well, `cube.intersection`) actually does the right thing and preserves both cell measures and ancil vars. Without that `and not` we'd end up with duplicate cell measures and ancil variables

schlunma replied:
I did not suggest to remove the `and not region_subset.cell_measures()`, but rather the check for `ndim == 1` 😁

valeriupredoi commented on esmvalcore/preprocessor/_area.py:
```diff
@@ -87,6 +91,66 @@ def extract_region(cube, start_longitude, end_longitude, start_latitude,
+    region_cell_meas = region_subset.cell_measures()
+    if (cell_measures, cube.coord('latitude').ndim
+            == 1) and not region_cell_meas:
+        from ._supplementary_vars import add_cell_measure
+        for cell_measure in cell_measures:
+            logger.info(
+                "Workaround: putting back cell "
+                "measure %s in cube %s", cell_measure, region_subset)
+            measure_name = cell_measure.measure
+            coord_spec = [(cube.coord("latitude"), 0),
+                          (cube.coord("longitude"), 1)]
+            cell_measure = iris.cube.Cube(
+                cell_measure.data,
+                var_name=cell_measure.var_name,
+                standard_name=cell_measure.standard_name,
+                dim_coords_and_dims=coord_spec,
+            )
+            cell_measure_subset = cell_measure.intersection(
+                longitude=(start_longitude, end_longitude),
+                latitude=(start_latitude, end_latitude),
+                ignore_bounds=True,
+            )
+            cell_measure_subset = cell_measure_subset.intersection(
+                longitude=(0., 360.))
+            add_cell_measure(region_subset, cell_measure_subset, measure_name)
```
I like - the idea but I'd stay away from that for now, Manu - only Loki knows what dragon we wake up in the case of irregular grids - like you see above, `cube.intersection` actually works *sometimes* (depending on the region's limits)

schlunma replied:
I don't see any problem with this...we are converting the cell measures to a cube here, on which we call `extract_region` (so basically the same `cube.intersection` code you have for 1D coordinates). I think we can and should be confident that `extract_region` works properly on cubes. After that, the cube is converted back to a cell measure and add to the actual data cube.

Also, the current implementation only works if the cell measures is 2D with dimensions lat/lon, whereas the suggestion here should work for any dimenion (e.g., cell volumes with an additional time and/or depth coordinate).

valeriupredoi replied:
Ok, cheers, Manu - I'll plug that in, progress vs being conservative - good point about that longitude and latitude strs, meant to change those but I overlooked them in the end, good catch 🦅

valeriupredoi replied:
unfortunately this approach makes the test run forever, I think it's the tuple generation that's serialized slower than a slow train

valeriupredoi replied:
or the cube indexing on a fairly heft tuple

valeriupredoi replied:
fixed it - Manu, you created the perfect infinite recursion there, calling extract_region from extract_region :rofl:

schlunma replied:
Ah damn, you're right, weneed to remove the `cell_measures` and `ancillary_variables` from the `cell_measure_cube ` before calling `extract_region` on it (the point of this was to actually call `extract_region` from within `extract_region` to make use of it's magic 😁). That way we should not end up in that infinite recursion.

E.g.,

```py
for sub_cm in cell_measure_cube.cell_measures():
    cell_measure_cube.remove_cell_measure(sub_cm)
for sub_av in cell_measure_cube.ancillary_variables():
    cell_measure_cube.remove_ancillary_variable(sub_av)
```

valeriupredoi replied:
this is a bit of a turkey, need a 5min break then I'll fix it

valeriupredoi commented on esmvalcore/preprocessor/_area.py:
```diff
@@ -467,15 +531,13 @@ def _process_ids(geometries, ids: list | dict | None) -> tuple:
                 f"If `ids` is given as dict, it needs exactly one entry, got "
-                f"{ids}"
-            )
+                f"{ids}")
```
ah yes, totally my bad, I yapf-ed it, flake8 was getting on me nerves - but, we don't use `black` so why the bother?

valeriupredoi replied:
if we use it for precommit stuff, I'd turn it off TBF, `black` is black magic and introduces a lot of shtuff on its own

schlunma replied:
> my bad for yapf-ing it, should I restore the old formatting from unrelated sections?

Yeah, would be nice to change it back 😁

valeriupredoi replied:
Sounds good, bud, shall revert, in a jiffy

valeriupredoi replied:
will tacke the other suggestions that I got from you (git them in already, but need to make sure all's shipshape, fix the tests too) then will backtrack on this, :bear: with me for a few, bud

valeriupredoi committed:
Manu suggestion

Co-authored-by: Manuel Schlund <32543114+schlunma@users.noreply.github.com>

valeriupredoi committed:
Manu suggestion to deal with irregular grids too

Co-authored-by: Manuel Schlund <32543114+schlunma@users.noreply.github.com>

valeriupredoi committed:
fix flake

valeriupredoi committed:
fix booboo from Manu

valeriupredoi committed:
plug in Manu's infinite recursion to fix it in terminal

Co-authored-by: Manuel Schlund <32543114+schlunma@users.noreply.github.com>

valeriupredoi committed:
unfubar the workflow

schlunma commented on esmvalcore/preprocessor/_area.py:
```diff
@@ -87,6 +91,54 @@ def extract_region(cube, start_longitude, end_longitude, start_latitude,
+                else 0 for d in range(cube.ndim)
+            ))
+            cell_measure_cube = cube[idx].copy(cell_measure.core_data())
+            cell_measure_cube_subset = cell_measure_cube.intersection(
+                longitude=(start_longitude, end_longitude),
+                latitude=(start_latitude, end_latitude),
+                ignore_bounds=True,
+            )
```
```suggestion
            for sub_cm in cell_measure_cube.cell_measures():
                cell_measure_cube.remove_cell_measure(sub_cm)
            for sub_av in cell_measure_cube.ancillary_variables():
                cell_measure_cube.remove_ancillary_variable(sub_av)
            cell_measure_cube_subset = extract_region(
                cell_measure_cube,
                start_longitude,
                end_longitude,
                start_latitude,
                end_latitude,
            )
```

schlunma commented on esmvalcore/preprocessor/_area.py:
```diff
@@ -87,6 +91,54 @@ def extract_region(cube, start_longitude, end_longitude, start_latitude,
+                slice(None) if d in ancil_var_dims else 0 for d in range(cube.ndim)
+            ))
+            ancil_var_cube = cube[idx].copy(ancil_var.core_data())
+            ancil_var_cube_subset = ancil_var_cube.intersection(
+                longitude=(start_longitude, end_longitude),
+                latitude=(start_latitude, end_latitude),
+                ignore_bounds=True,
+            )
```
```suggestion
            for sub_cm in ancil_var_cube.cell_measures():
                ancil_var_cube.remove_cell_measure(sub_cm)
            for sub_av in ancil_var_cube.ancillary_variables():
                ancil_var_cube.remove_ancillary_variable(sub_av)
            ancil_var_cube_subset = extract_region(
                ancil_var_cube,
                start_longitude,
                end_longitude,
                start_latitude,
                end_latitude,
            )
```

valeriupredoi committed:
Update esmvalcore/preprocessor/_area.py

Co-authored-by: Manuel Schlund <32543114+schlunma@users.noreply.github.com>

valeriupredoi committed:
Update esmvalcore/preprocessor/_area.py

Co-authored-by: Manuel Schlund <32543114+schlunma@users.noreply.github.com>

valeriupredoi commented:
with the latest changes when you try print the cube that's been changed you get this:
```
Traceback (most recent call last):
  File "/home/valeriu/test_intersect.py", line 27, in <module>
    print("ex1", ex1)
  File "/home/valeriu/miniconda3/envs/esmvaltool/lib/python3.11/site-packages/iris/cube.py", line 2563, in __str__
    return self.summary()
           ^^^^^^^^^^^^^^
  File "/home/valeriu/miniconda3/envs/esmvaltool/lib/python3.11/site-packages/iris/cube.py", line 2558, in summary
    printer = CubePrinter(self)
              ^^^^^^^^^^^^^^^^^
  File "/home/valeriu/miniconda3/envs/esmvaltool/lib/python3.11/site-packages/iris/_representation/cube_printout.py", line 165, in __init__
    cube_summary = CubeSummary(cube_or_summary)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/valeriu/miniconda3/envs/esmvaltool/lib/python3.11/site-packages/iris/_representation/cube_summary.py", line 364, in __init__
    add_vector_section(
  File "/home/valeriu/miniconda3/envs/esmvaltool/lib/python3.11/site-packages/iris/_representation/cube_summary.py", line 355, in add_vector_section
    self.vector_sections[title] = VectorSection(
                                  ^^^^^^^^^^^^^^
  File "/home/valeriu/miniconda3/envs/esmvaltool/lib/python3.11/site-packages/iris/_representation/cube_summary.py", line 203, in __init__
    self.contents = [
                    ^
  File "/home/valeriu/miniconda3/envs/esmvaltool/lib/python3.11/site-packages/iris/_representation/cube_summary.py", line 204, in <listcomp>
    VectorSummary(cube, vector, iscoord) for vector in vectors
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/valeriu/miniconda3/envs/esmvaltool/lib/python3.11/site-packages/iris/_representation/cube_summary.py", line 141, in __init__
    dims = vector.cube_dims(cube)
           ^^^^^^^^^^^^^^^^
AttributeError: 'Cube' object has no attribute 'cube_dims'
```

valeriupredoi commented:
the test is failing as well bc the metadata of the cube with ancil vars in preserves the metadata from the original cube `AssertionError: assert 'sftlf' == 'psl'`

valeriupredoi commented:
am going back to the the latest working commit 93b085871e9a076156fd9a2fac16408f6b556e59

valeriupredoi committed:
Revert "Update esmvalcore/preprocessor/_area.py"

This reverts commit bf35614a850ae4db30623d65ecf04cd57c585b56.

valeriupredoi committed:
Revert "Update esmvalcore/preprocessor/_area.py"

This reverts commit 74b979bb1c246a22e5d2cb2975cb7f6f1e6f6c57.

valeriupredoi committed:
Revert "unfubar the workflow"

This reverts commit 2efbc9e29df6dd2a347c3346aa08ea86e019a2e4.

valeriupredoi committed:
Revert "plug in Manu's infinite recursion to fix it in terminal"

This reverts commit 3f0dede551d4edbd131531f027b6823ebdb40353.

valeriupredoi committed:
Revert "fix booboo from Manu"

This reverts commit 85e8b33b64e1ea778726f15f4eeeb17a8732c9d7.

valeriupredoi committed:
Revert "fix flake"

This reverts commit 429a0c7f6054692b7a54a1e6f1483fa266fa166a.

valeriupredoi committed:
Revert "Manu suggestion to deal with irregular grids too"

This reverts commit 6d8652a2f389aef6ce0c1d2e10ab8c0721769adf.

valeriupredoi committed:
Revert "Manu suggestion"

This reverts commit 15d497348e3af975e1463ae0bac33545d94e3677.

valeriupredoi committed:
apply Manu suggestion to generalize for irreg grids too

valeriupredoi commented:
sorry @schlunma a few too many issues with your review suggestions (both at our end and iris - iris issue which I am not prepared to dive into at the moment, see https://github.com/ESMValGroup/ESMValCore/pull/2166#issuecomment-1669576532); if you think we'd benefit enough from those changes, I suggest we merge this, then you can open a PR based off `main` where you apply those changes, and test properly. I'll fix the test not to use the test data (and delete that file), then I'll revert the style changes resulted from `yapf`ing it, then I am ready to merge this :beer:

valeriupredoi committed:
re-black code

valeriupredoi committed:
missed a couple blacks

valeriupredoi committed:
optimized test without need of data file

valeriupredoi committed:
rm data file

valeriupredoi commented:
OK Manu, I have finished poslishing this:
- used your suggestion to allow for irregular grids to be looked at
- fixed the test and removed the now unnecessary test data file
- re-blacked `_area.py` (well, reverted previously added style changes from yapf)

I think we should do this: merge this to have correct functionality for extract_region, then you and me we can go back to the suggestions you made earlier in a follow up PR, unfortunately there are quite a few hitches (not least the iris one that we should look at) to be looked at here; we may as well not even do it all if we'll remove all this when iris fix their intersection, but at least I'd like to understand that print(cube) issue above. What say you? :beer:

valeriupredoi commented:
Also, as a note to all of us (I am guilty as charged myself) - everytime we suggest major code changes in a PR, in the form of suggested code via the browser/GitHub API, it would be a lot easier for the PR owner if the reviewer tested those changes by either running the test provided in the PR or by running a small independent test - this, to catch serious issues (not style, minor issues with the suggestions, those can be taken care of the PR owner, after including the suggested code changes). I think we should add this to our best practices PR reviewing docs :+1:

schlunma commented:
Well, if you use `cube.intersection` (and not `extract_region`) to get the cell measure sub cube, this won't work for irregular grids. Hence you need to reinstate the check for `coord('latitude').ndim == 1`.

Also, in this form, this won't work (and will actually crash) for cell measures and ancillary variables that do not exactly span latitude/longitude. We need to take care of this properly here, otherwise it will break recipes.

I am honestly confused why my suggestion didn't work. From the failing test on CircleCI it looks just like a `flake8` [fail](https://app.circleci.com/pipelines/github/ESMValGroup/ESMValCore/9518/workflows/fedbbda2-6e1d-4b49-a08f-1ea8be52b428/jobs/41505)? I will give it a try myself in a local branch.

valeriupredoi commented:
> Well, if you use `cube.intersection` (and not `extract_region`) to get the cell measure sub cube, this won't work for irregular grids. Hence you need to reinstate the check for `coord('latitude').ndim == 1`.
> 
> Also, in this form, this won't work (and will actually crash) for cell measures and ancillary variables that do not exactly span latitude/longitude. We need to take care of this properly here, otherwise it will break recipes.
> 
> I am honestly confused why my suggestion didn't work. From the failing test on CircleCI it looks just like a `flake8` [fail](https://app.circleci.com/pipelines/github/ESMValGroup/ESMValCore/9518/workflows/fedbbda2-6e1d-4b49-a08f-1ea8be52b428/jobs/41505)? I will give it a try myself in a local branch.

sounds like a very good plan! Cheers, Manu! Can grab this branch, edit locally, and when things look OK to you just push changes here - that was my point in https://github.com/ESMValGroup/ESMValCore/pull/2166#issuecomment-1669748583 - reviewer's minimal testing is something we should start doing :beer:

schlunma commented:
Tests pass fine for me locally (your test and my suggestion). Can I push it? 😁

valeriupredoi commented:
> Tests pass fine for me locally (your test and my suggestion). Can I push it? grin

Pliz do :beer:

valeriupredoi committed:
Merge branch 'main' into readd_cell_measure_after_extract_region

schlunma committed:
Made extract_region work with cell measures and ancillaries with every dimensionality

valeriupredoi commented:
@schlunma let's get this in plz so I can help @rebeccaherman1 with an install at DKRZ :beer:

schlunma reviewed (approved):
Thanks V! Can I be a bother and ask to to test this with the script mentioned [here](https://github.com/ESMValGroup/ESMValCore/issues/2162#issuecomment-1664161639)? I think we need to make sure that this properly works for cubes with time dimensions, too (If you like, you could also add a brief test of this 😁).

I will be in a seminar today with limitied availability, so I pre-approve this. Thanks for the fix!!

valeriupredoi commented:
> Thanks V! Can I be a bother and ask to to test this with the script mentioned [here](https://github.com/ESMValGroup/ESMValCore/issues/2162#issuecomment-1664161639)? I think we need to make sure that this properly works for cubes with time dimensions, too (If you like, you could also add a brief test of this grin).
> 
> I will be in a seminar today with limitied availability, so I pre-approve this. Thanks for the fix!!

Very many thanks, Manu! Yes, that's what I've been using to test the PR (alas, slightly changed, but I'll test with the exact script that Rebecca had initially). Good point, I'll put a time dim in the integration test in this PR :beer:

valeriupredoi committed:
add time dim to test cube

valeriupredoi committed:
fix indentation
